File: src/main/java/com/everyonewaiter/global/logging/MDCLoggingFilter.java
package com.everyonewaiter.global.logging;

import static com.everyonewaiter.global.logging.HttpRequestParser.parseHeaders;
import static com.everyonewaiter.global.logging.HttpRequestParser.parseParameters;
import static com.everyonewaiter.global.logging.HttpRequestParser.parseRequestURI;
import static com.everyonewaiter.global.logging.HttpRequestParser.parseXRequestId;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.io.IOException;
import lombok.RequiredArgsConstructor;
import org.slf4j.MDC;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

@Component
@RequiredArgsConstructor
class MDCLoggingFilter extends OncePerRequestFilter {

  private final ObjectMapper objectMapper;

  @Override
  protected void doFilterInternal(
      @NonNull HttpServletRequest request,
      @NonNull HttpServletResponse response,
      @NonNull FilterChain filterChain
  ) throws ServletException, IOException {
    MDC.put("requestId", parseXRequestId(request));
    MDC.put("requestMethod", request.getMethod());
    MDC.put("requestURI", parseRequestURI(request));
    MDC.put("requestParameters", parseParameters(request));
    MDC.put("requestHeaders", parseHeaders(request));
    MDC.put("requestCookies", objectMapper.writeValueAsString(request.getCookies()));

    filterChain.doFilter(request, response);
  }

}

File: src/main/kotlin/com/everyonewaiter/global/filter/MDCLoggingFilter.kt
package com.everyonewaiter.global.filter

import com.everyonewaiter.global.extension.headers
import com.everyonewaiter.global.extension.parameters
import com.everyonewaiter.global.extension.requestUri
import com.everyonewaiter.global.extension.xRequestId
import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.MDC
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
class MDCLoggingFilter(
    private val objectMapper: ObjectMapper,
) : OncePerRequestFilter() {
    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        MDC.put("requestId", request.xRequestId)
        MDC.put("requestUri", request.requestUri)
        MDC.put("requestParameters", request.parameters)
        MDC.put("requestHeaders", request.headers)
        MDC.put("requestCookies", objectMapper.writeValueAsString(request.cookies))
        filterChain.doFilter(request, response)
    }
}
